Add man page for condor_token_*. #7361 (#83)
htcondor/htcondor: docs/man-pages/condor_token_create.rst

      

*condor_token_create*
======================

> given a password file, create an authentication token

Reviewer: ```suggestion
given a pool password file to create an authentication token for the TOKEN auth method
```
or some other way of making it explicit that this is (only) for TOKEN auth.

:index:`condor_token_create<single: condor_token_create; HTCondor commands>`\ :index:`condor_token_create command`

Synopsis
--------

**condor_token_create** **-identity** *user@domain* [**-keyid** *keyid*]
[**-authz** *authz* ...] [**-lifetime** *value*]
[**-token** *filename*] [**-debug**]

**condor_token_create** [**-help** ]

Description
-----------

*condor_token_create* will read a HTCondor password file inside the

> *SEC_PASSWORD_DIRECTORY* and use it to create an authentication token.

Reviewer: Add "by default, this is the pool password file" or something to that effect; it's kind of buried otherwise.

The authentication token may be subsequently used by clients to authenticate

> against a remote HTCondor server.  Tokens allow fine-grained authentication

Reviewer: How about "Tokens allow fine-grained authentication as individual HTCondor users, as opposed to pool password, where anything in possession of the pool password will authenticate as the same user." ?

Reviewer: Alternatively, the rationale for using TOKEN instead of PASSWORD can just be a link to the [token section of the manual](https://htcondor.readthedocs.io/en/latest/admin-manual/security.html#token-authentication).

Author: I'm a bit loath to put links into a man page; I'll clarify the text inline.

 **-debug**
    Causes debugging information to be sent to ``stderr``, based on the
    value of the configuration variable ``TOOL_DEBUG``.
 **-help**
    Display brief usage information and exit.
 **-identity** *user@domain*
    Set a specific client identity to be written into the token; a client
    will authenticate as this identity with a remote server.
 **-keyid** *keyid*
    Specify a specific key to use in the directory specified by the
    *SEC_PASSWORD_DIRECTORY* configuration variable. The key name must
    match a file in the password directory; the file's contents must
    be created with *condor_store_cred* and will be used to sign the
    resulting token.  If **-keyid** is not set, then the default pool
    password will be used.
 **-lifetime** *value*
    Specify the lifetime, in seconds, for the token to be valid (the
    token validity will start when the token is signed).  After the
    lifetime expires, the token cannot be used for authentication.  If
    not specified, the token will contain no lifetime restrictions.
 **-token** *filename*
    Specifies a filename, relative to the directory in the *SEC_TOKEN_DIRECTORY*

>     configuration variable (defaulting to ``~/.condor/tokens.d``), where

Reviewer: Where is this on Windows?

Author: I have no clue; I'll turn this into an example instead of authoritative.

      

*condor_token_create*
======================

given a password file, create an authentication token for the TOKEN authentication method
:index:`condor_token_create<single: condor_token_create; HTCondor commands>`\ :index:`condor_token_create command`

Synopsis
--------

**condor_token_create** **-identity** *user@domain* [**-keyid** *keyid*]
[**-authz** *authz* ...] [**-lifetime** *value*]
[**-token** *filename*] [**-debug**]

**condor_token_create** [**-help** ]

Description
-----------

*condor_token_create* will read an HTCondor password file inside the

> *SEC_PASSWORD_DIRECTORY* (by default, this is the pool password) and use it to create an authentication token.

Author: Added a comment in the header noting this is only expected to work on Unix right now (apologies!)

The authentication token may be subsequently used by clients to authenticate
against a remote HTCondor server.  Tokens allow fine-grained authentication
as individual HTCondor users as opposed to pool password, where anything
in possession of the pool password will authenticate as the same user.

An identity must be specified for the token; this will be the client's
resulting identity at the remote HTCondor server.
If the **-lifetime** or (one or more) **-authz** options are specified,
the token will contain additional restrictions that limit what the
client will be authorized to do.

If successful, the resulting token will be sent to ``stdout``; by specifying

> the **-token** option, it will instead be written to the user's token directory.

Reviewer 2: 1) That would be cool to add information about question how to make a token for the daemon-to-daemon authentication. How can we use `SEC_TOKEN_SYSTEM_DIRECTORY`? In the case of Windows?

2) Suppose that on the first machine (Unix with master, collector, negotiator and schedd), we generated a token for `tester@windows-qa`. On the second `windows-qa` Windows machine, a STARTD demon is deployed under the user `tester` (administrator). The first Unix machine does not see the second (it is behind the firewall), a CCB mechanism is used.
The user wants the EXECUTE node to be added to the pool. Do the user need to transfer token to the Windows machine? Where to put this token?

3) I would like to briefly learn the following from this document: how can these tokens be used and why should they be written to files? And why should these files be hidden from attackers?

Author: @hozblok - thanks for the feedback!

1.  If a compatible token is present in the system's token directory, it will be used from daemon-to-daemon.  Apologies - I do not know anything about how to set this up on Windows (would love to hear the answers to your questions though!).  That might be a good question for the HTCondor user's list.  @johnkn - have you tried the TOKEN workflow on Windows?
2.  If created with `condor_token_create`, the user could transfer the token to the Windows machine by hand (scp, save it on USB, print it out), install it in the appropriate directory, and enable TOKEN auth (default in 8.9.4).  _However_, rather than potentially-insecurely transferring the token, it is better to use `condor_token_request` (or use the fact the startd should automatically request tokens) from `windows-qa` and have the admin approve the token request.  The `condor_token_request` CLI will securely transfer things.
3.  I think the full security model is probably a bit much for the manpages.  I will try to address as much as I can in the next revision.

Great input!

Author: I think I addressed these as best I can.  Can you take a look?

Reviewer 2: @bbockelm Thanks! It became much better. Special thanks for the detailed answers, I removed some of the gaps in my understanding and.. I have sent additional questions to the HTCondor user's list.

`I think the full security model is probably a bit much for the manpages.` I agree with you.

Options
-------

 **-authz** *authz*
    Adds a restriction to the token so it is only valid to be used for
    a given authorization level (such as ``READ``, ``WRITE``, ``DAEMON``,
    ``ADVERTISE_STARTD``).  If multiple authorizations are needed, then
    **-authz** must be specified multiple times.  If **-authz** is not
    specified, no authorization restrictions are added and authorization
    will be solely based on the token's identity.
    **NOTE** that **-authz** cannot be used to give an identity additional
    permissions at the remote host.  If the server's admin only permits
    the user ``READ`` authorization, then specifying ``-authz WRITE`` in a
    token will not allow the user to perform writes.
 **-debug**
    Causes debugging information to be sent to ``stderr``, based on the
    value of the configuration variable ``TOOL_DEBUG``.
 **-help**
    Display brief usage information and exit.
 **-identity** *user@domain*
    Set a specific client identity to be written into the token; a client
    will authenticate as this identity with a remote server.
 **-keyid** *keyid*

>     Specify a specific key to use in the directory specified by the

Reviewer 2: That would be great to add example with `condor_store_cred` and `-keyid`

>     Specify a specific key to use in the directory specified by the

Author: Added an improved example in the latest commit.

>     Specify a specific key to use in the directory specified by the

Reviewer 2: Great example, I was waiting for this!

[**-token** *filename*] [**-debug**]

**condor_token_create** [**-help** ]

Description
-----------

*condor_token_create* will read an HTCondor password file inside the
*SEC_PASSWORD_DIRECTORY* (by default, this is the pool password) and use it to create an authentication token.
The authentication token may be subsequently used by clients to authenticate
against a remote HTCondor server.  Tokens allow fine-grained authentication
as individual HTCondor users as opposed to pool password, where anything
in possession of the pool password will authenticate as the same user.

An identity must be specified for the token; this will be the client's
resulting identity at the remote HTCondor server.
If the **-lifetime** or (one or more) **-authz** options are specified,
the token will contain additional restrictions that limit what the
client will be authorized to do.
If an attacker is able to access the token, they will be able to authenticate
with the identity listed in the token (subject to the restrictions above).

If successful, the resulting token will be sent to ``stdout``; by specifying
the **-token** option, it will instead be written to the user's token directory.

> If written to to *SEC_TOKEN_SYSTEM_DIRECTORY* (default ``/etc/condor/tokens.d``),

Reviewer 2: Excellent, but there is a small typo: double `to`

> If written to to *SEC_TOKEN_SYSTEM_DIRECTORY* (default ``/etc/condor/tokens.d``),

Reviewer: ```suggestion
If written to *SEC_TOKEN_SYSTEM_DIRECTORY* (default ``/etc/condor/tokens.d``),
```
(thanks @hozblok)

*condor_token_create* is only currently supported on Unix platforms.

Options
-------

 **-authz** *authz*
    Adds a restriction to the token so it is only valid to be used for
    a given authorization level (such as ``READ``, ``WRITE``, ``DAEMON``,
    ``ADVERTISE_STARTD``).  If multiple authorizations are needed, then
    **-authz** must be specified multiple times.  If **-authz** is not
    specified, no authorization restrictions are added and authorization
    will be solely based on the token's identity.
    **NOTE** that **-authz** cannot be used to give an identity additional
    permissions at the remote host.  If the server's admin only permits
    the user ``READ`` authorization, then specifying ``-authz WRITE`` in a
    token will not allow the user to perform writes.
 **-debug**
    Causes debugging information to be sent to ``stderr``, based on the
    value of the configuration variable ``TOOL_DEBUG``.
 **-help**
    Display brief usage information and exit.
 **-identity** *user@domain*
    Set a specific client identity to be written into the token; a client
    will authenticate as this identity with a remote server.
 **-key** *keyid*

>     Specify a specific key to use in the directory specified by the

Reviewer: ```suggestion
    Specify a key file to use under the directory specified by the
```

    *SEC_PASSWORD_DIRECTORY* configuration variable. The key name must
    match a file in the password directory; the file's contents must
    be created with *condor_store_cred* and will be used to sign the

>     resulting token.  If **-keyid** is not set, then the default pool

Reviewer: ```suggestion
    resulting token.  If **-key** is not set, then the default pool
```
